runtime/event_loop.cpp

```diff
@@ -38,40 +40,63 @@ bool EventLoop::cancel_async_task(api::Engine *engine, const int32_t id) {
 
 bool EventLoop::has_pending_async_tasks() { return !queue.get().tasks.empty(); }
 
-bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
-  // Loop until no more resolved promises or backend requests are pending.
-  // LOG("Start processing async jobs ...\n");
+void EventLoop::incr_event_loop_interest() { queue.get().interest_cnt++; }
+
+void EventLoop::decr_event_loop_interest() {
+  MOZ_ASSERT(queue.get().interest_cnt > 0);
+  queue.get().interest_cnt--;
+}
+
+inline bool interest_complete() { return queue.get().interest_cnt == 0; }
 
+inline void exit_event_loop() { queue.get().event_loop_running = false; }
+
+bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
+  if (queue.get().event_loop_running) {
+    fprintf(stderr, "cannot run event loop as it is already running");
+    return false;
+  }
+  queue.get().event_loop_running = true;
   JSContext *cx = engine->cx();
 
-  do {
-    // First, drain the promise reactions queue.
-    while (js::HasJobsPending(cx)) {
-      js::RunJobs(cx);
+  while (true) {
+    // Run a microtask checkpoint
+    js::RunJobs(cx);
 
-      if (JS_IsExceptionPending(cx))
-        return false;
+    if (JS_IsExceptionPending(cx)) {
+      exit_event_loop();
+      return false;
+    }
+    if (interest_complete()) {
+      exit_event_loop();
+      return true;
     }
 
-    // TODO: add general mechanism for extending the event loop duration.
-    // Then, check if the fetch event is still active, i.e. had pending promises
-    // added to it using `respondWith` or `waitUntil`.
-    // if (!builtins::FetchEvent::is_active(fetch_event))
-    //   break;
-
-    // Process async tasks.
-    if (has_pending_async_tasks()) {
-      auto tasks = &queue.get().tasks;
-      const auto index = api::AsyncTask::select(tasks);
-      auto task = tasks->at(index);
-      if (!task->run(engine)) {
+    const auto tasks = &queue.get().tasks;
+    size_t tasks_size = tasks->size();
+    if (tasks_size == 0) {
+      exit_event_loop();
+      // if there is no interest in the event loop at all, we always run one tick
+      if (interest_complete()) {
+        return true;
+      } else {
+        fprintf(stderr, "event loop error - both task and job queues are empty, but expected "
+                        "operations did not resolve");
         return false;
       }
-      tasks->erase(tasks->begin() + index);
     }
-  } while (js::HasJobsPending(engine->cx()) || has_pending_async_tasks());
 
-  return true;
+    // Select the next task to run according to event-loop semantics of oldest-first.
+    size_t task_idx = api::AsyncTask::select(tasks);
```

Comment:
One of the things that can be done here so this doesn't block indefinitely if an I/O task never finishes, is to create a timer task if there are pending CPU-only things -- and make the event loop work only with `api::AsyncTask::select()`.

The timer task can be something like akin to a `setTimeout(runMicroTaskCheckpoints, 0)` if there are any microtasks to be executed.

This way it's easier to integrate pollable/selectable tasks with things that are purely CPU-bound, even if your `select()` primitive doesn't offer a timeout parameter.

Reply:
(This would be akin to GLib's [`g_idle_add()`](https://docs.gtk.org/glib/func.idle_add.html).  In fact, it might be a good idea to read the documentation for the GLib main loop, as it's a tried and true design.)

Reply:
Unfortunately if I understand the proposal correctly that wouldn't be in compliance with how the DOM event loop works, which is what we're aiming for here. As mentioned in my last review, we always have to run the microtask checkpoint after every task, instead of enqueuing microtask handling for a later time.